examples/gateway/car-file-gateway/api.go

```diff
@@ -0,0 +1,234 @@
+	"github.com/ipfs/interface-go-ipfs-core/options"
+	"github.com/ipfs/interface-go-ipfs-core/path"
+	carblockstore "github.com/ipld/go-car/v2/blockstore"
+	dagpb "github.com/ipld/go-codec-dagpb"
+	ipldPrime "github.com/ipld/go-ipld-prime"
+	"github.com/ipld/go-ipld-prime/node/basicnode"
+	"github.com/ipld/go-ipld-prime/schema"
+)
+
+type carGateway struct {
+	bs   blockstore.Blockstore
+	bsrv blockservice.BlockService
+	dsrv ipld.DAGService
+}
+
+func newCarGateway(r io.ReaderAt) (*carGateway, error) {
+	bs, err := carblockstore.NewReadOnly(r, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	bsrv := blockservice.New(bs, offline.Exchange(bs))
+	dsrv := merkledag.NewDAGService(bsrv)
+
+	return &carGateway{
+		bs:   bs,
+		bsrv: bsrv,
+		dsrv: dsrv,
+	}, nil
+}
+
+func (api *carGateway) ResolvePath(ctx context.Context, p path.Path) (path.Resolved, error) {
+	if _, ok := p.(path.Resolved); ok {
+		return p.(path.Resolved), nil
+	}
+
+	if err := p.IsValid(); err != nil {
+		return nil, err
+	}
+
+	if p.Namespace() != "ipfs" {
+		return nil, fmt.Errorf("unsupported path namespace: %s", p.Namespace())
+	}
+
+	ipath := ipfspath.Path(p.String())
+
+	ipldFetcher := bsfetcher.NewFetcherConfig(api.bsrv)
+	ipldFetcher.PrototypeChooser = dagpb.AddSupportToChooser(func(lnk ipldPrime.Link, lnkCtx ipldPrime.LinkContext) (ipldPrime.NodePrototype, error) {
+		if tlnkNd, ok := lnkCtx.LinkNode.(schema.TypedLinkNode); ok {
+			return tlnkNd.LinkTargetNodePrototype(), nil
+		}
+		return basicnode.Prototype.Any, nil
+	})
+	fetcher := ipldFetcher.WithReifier(unixfsnode.Reify)
+	resolver := ipfspathresolver.NewBasicResolver(fetcher)
+
+	node, rest, err := resolver.ResolveToLastNode(ctx, ipath)
+	if err != nil {
+		return nil, err
+	}
+
+	root, err := cid.Parse(ipath.Segments()[1])
+	if err != nil {
+		return nil, err
+	}
+
+	return path.NewResolvedPath(ipath, node, root, gopath.Join(rest...)), nil
+}
+
+func (api *carGateway) resolveNode(ctx context.Context, p path.Path) (ipld.Node, error) {
+	rp, err := api.ResolvePath(ctx, p)
+	if err != nil {
+		return nil, err
+	}
+
+	node, err := api.dsrv.Get(ctx, rp.Cid())
+	if err != nil {
+		return nil, fmt.Errorf("get node: %w", err)
+	}
+	return node, nil
+}
+
+func (api *carGateway) GetUnixFsNode(ctx context.Context, p path.Resolved) (files.Node, error) {
+	nd, err := api.resolveNode(ctx, p)
+	if err != nil {
+		return nil, err
+	}
+
+	return unixfile.NewUnixfsFile(ctx, api.dsrv, nd)
+}
+
+func (api *carGateway) LsUnixFsDir(ctx context.Context, p path.Resolved) (<-chan iface.DirEntry, error) {
+	settings, err := options.UnixfsLsOptions(
+		options.Unixfs.ResolveChildren(false),
+		options.Unixfs.UseCumulativeSize(true),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	dagnode, err := api.resolveNode(ctx, p)
+	if err != nil {
+		return nil, err
+	}
+
+	dir, err := uio.NewDirectoryFromNode(api.dsrv, dagnode)
+	if err == uio.ErrNotADir {
+		return api.lsFromLinks(ctx, dagnode.Links(), settings)
+	}
```

Comment:
This is `LsUnixFsDir` we can't get a `uio.ErrNotADir`, right?

Reply:
@aschmahmann that's a good question. I traced it back to this commit from 5 years ago: https://github.com/ipfs/kubo/commit/a22cae1bce7dfa9b67ed5c969c469e259eac0fd9 - unfortunately it is not connected to any PR so I can get some more context. I removed it locally and it works well with the directory I am testing with. I wonder if there's some old format that led to this decision with two different types. Do you know anything about it?

Reply:
As per our other convo, this seems to be related to `ipfs ls` also supporting listing the links of files: https://github.com/ipfs/kubo/issues/9541